### ISSUE TITLE
UnauthenticatedException and UnauthorizedException are mixed up

### ISSUE THREAD

**Issue by ghost** [makumba/makumba#204]
**Reported by @frederikhabils on 22 May 2003 14:49 UTC**
bug 443 comment 4 reminded me of some concern I've been feeling about 
UnauthenticatedException and UnauthorizedException not being used in the 
correct way, in applications.

I believe the commonly-used definition is 

Unauthenticated
  = you did not yet identify yourself towards the system
Unauthorized
  = (you did authenticate yourself, but) you don't have the right to do that

Now, the thing is that the UnauthenticatedException prints (thru mak:response) 
some 'crappy' message like "could not find ..." along with the text added by 
the BL programmer. UnauthorizedException does not print that mesg, and only 
prints the thing that the BL programmer passes in the "throw new UnauthExc(msg)"

Apart from that message, both Exceptions provoke the same action: displaying 
login.jsp

I've gotten the impression that for this reason, at many places in Karamba, 
UnauthorizedException is being used, where UnauthenticatedException is 
appropriate.

Why is this a problem?

I believe that the response to UnauthenticatedExc should be login.jsp
The response to UnauthorizedExc should not be login.jsp, but should be sorry-
you-cant-do-that-go-back-and-do-something-else.jsp

Step by step:

Suppose the logged in poorUser tries to do an adminAction (e.g. edit a 
group)... then BL throws an UnauthorizedException and the poorUser currently 
sees the login page. 

Now, if nobody were logged in, that login page will try to set e.g. session 
attribute 'member'. 

But somebody already being logged in, the result of the login page submit (IMO) 
will always be that the same member remains logged in (even if the user/pass of 
some adminUser were enter on the login)... since 
   checkAttributes(){
      a.getAttribute("member");
   }
-> will find the session attribute "member", thus it will not do the findMember
() function, which could allow the adminUser to log in.

Basically, in the current design it is (I think) a dead end solution to show 
login.jsp to an UnauthorizedException. And programmers have used 
UnauthorizedException where UnauthenticatedException was the way to go.

Migrated-From: http://trac.makumba.org/ticket/444

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 22 May 2003 15:41 UTC**
agreed. if there is an authentication for that secret variable and that value of
the secret var is not authorized, it means that there are no rights, and the
respective individual should not have a chance to try to login in the name of
somebody else.

should karamba BL be fixed for this? i think that a message can be added to
every UnauthorizedException thrown, explaining in human-readable terms why there
is no access (you cannot do this because you need be an admin of the kind X or
Y, etc). the BL is the right place to compose such messages, because they are
part of the application logic. if UnauthorizedExc doesn't have that kind of
constructor, one should be made.

as for makumba, i think that the most elegant solution is
org_makumba_UnauthorizedException.jsp

**Comment by ghost** [comment]
**Comment by @frederikhabils on 22 May 2003 21:27 UTC**
comment 0:
> Now, the thing is that the UnauthenticatedException prints (thru 
> mak:response) some 'crappy' message like "could not find ..." along with the
> text added by 

Correction: the message from Unauth*Exception is not printed by mak:response.
It is found through this call:
   String msg=pageContext.findAttribute(pageContext.EXCEPTION).toString();

The "crappy" message I referred to (and that IMO has made mak-users use 
UnauthorizedExc), is added by 

  public class UnauthenticatedException extends UnauthorizedException
    public String getMessage() { return super.getMessage()+" could not set 
attribute "+attName; }


in response to comment 1:

- Yes, the Unauth*Exception error have an initialiser with String, and String 
is used by Karamba BL to explain the reason of not logging in.

- Yes, it would be good to review Karamba BL, to make sure that Unauthorized 
and Unauthenticated are used appropriately. But to do so, I suggest to first 
fix the unwanted ('crappy') message addition : " could not set 
attribute "+attName;

Btw, the difficult way of obtaining the Exception message, may indicate it was 
not from the beginning intended to end up in the web-GUI?

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 23 May 2003 07:12 UTC**
> The "crappy" message I referred to (and that IMO has made mak-users use 
> UnauthorizedExc), is added by 

yes, it seems that getMessage is sort of hijacked in UnauthenticatedException...
not a smart move from my egocentric person

> - Yes, it would be good to review Karamba BL, to make sure that Unauthorized 
> and Unauthenticated are used appropriately. But to do so, I suggest to first 
> fix the unwanted ('crappy') message addition : " could not set 
> attribute "+attName;

agreed, but don't wait for me to do it.

> Btw, the difficult way of obtaining the Exception message, may indicate it was 
> not from the beginning intended to end up in the web-GUI?

well, PageContext.EXCEPTION is the JSP-standard way of retrieving the exceptions
in JSP errroPage's. it is also accessible as the "exception" Java variable in
the 'true' JSP error pages. if we want to reduce scriplets (which we want),
we'll have to support 

<mak:attribute name="makumba.exception" /> 

or so, in all exception-handling support cases (as per bug 443)

somehow also the supplementary information of the exception (e.g. the attribute
name for the unauthenticated) should be retrievable. but we can leave that to
scriplet or (later) JSTL operations.

since we're talking about changing a number of Karamba sources (also in bug 41),
i think that a full regression test is also an opportunity to test those
changes. after all, it's karamba regression that we're generally interested in,
and whether some error is a makumba-specific regression or karamba-specific, we
can investigate on a case-by-case basis.

**Comment by ghost** [comment]
**Comment by @frederikhabils on 29 May 2003 13:09 UTC**
> agreed, but don't wait for me to do it.

well, I recursively updated fred-m, removed the message from 
org/makumba/UnauthenticatedException.java and committed from parade. 

very nice, but what should happen next? 

I'm getting a headache from the million cvs tags, branches, ant jobs, makumba 
versions, etc. that seem to be involved.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 29 May 2003 14:45 UTC**
> well, I recursively updated fred-m, removed the message from 
> org/makumba/UnauthenticatedException.java and committed from parade. 
 
> I'm getting a headache from the million cvs tags, branches, ant jobs, makumba 
> versions, etc. that seem to be involved.

you committed in the MAIN branch so you have no problem with branches

before committing you could have done ant getLocalMakumba in fred-k, and test a
bit (and make the necessary Unauthenticated <-> Unauthorized BL changes)

then you can simply publishJar a new 0.5.7.x version... we still don't have any
decission on the "CVS and releases" mail thread, so we do as before

but i'd guess the new makumba version should also be told not to show login.jsp
for unauthorized, right?

**Comment by ghost** [comment]
**Comment by @frederikhabils on 8 Sep 2003 21:07 UTC**
*** Bug 559 has been marked as a duplicate of this bug. ***

**Comment by ghost** [comment]
**Comment by @frederikhabils on 8 Sep 2003 21:14 UTC**
Clarification : 
the tiny improvement discussed in comment 4, is part of makumba since 0.5.9
http://cvs.makumba.org/cgi-bin/cvsweb.cgi/makumba/classes/org/makumba/UnauthenticatedException.java

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:22 UTC**

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:36 UTC**

**Comment by ghost** [comment]
**Comment by @manuelbernhardt on 29 Jun 2010 19:56 UTC**
I think rudi managed to display an OK error message for Unauthorized stuff in K!

But I wonder, how would that need to be handled by a usual mak application? i.e. what needs to be done & documented for this ticket to be closed?

**Comment by ghost** [comment]
**Comment by rudi on 30 Jun 2010 08:27 UTC**
I think #1151 covers this in an explicit manner, i.e. it proposes a solution to the main problem, while the rest of the ticket (talking about how it is used wrongly in K!) is not Makumba relevant. So this might be closed

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 1 Jul 2010 09:25 UTC**
